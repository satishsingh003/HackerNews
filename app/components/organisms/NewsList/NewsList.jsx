import React from 'react';
import PropTypes from 'prop-types';
import NewsListItem from '../NewsListItem';
import { PAGE_TITLE } from '../../../constants';

const NewsList = ({newsListData, loadMoreNews, upVoteNewsItem, hideNewsItem}) => {
    return (
        <>  
            <h1 className="hidden">{PAGE_TITLE}</h1>
            <ul>
                {newsListData.map(listItem => (
                    <NewsListItem 
                        key={listItem.objectID} 
                        newsItem={listItem} 
                        hideNewsItem={hideNewsItem} 
                        upVoteNewsItem={upVoteNewsItem} 
                    />
                ))}
            </ul>
            <a href="#" onClick={loadMoreNews} className="loadMore">Load More</a>
            <style jsx>{`
                .hidden {
                    position: absolute;
                    left: -999px;
                    visibility: hidden;
                }
                .loadMore {
                    text-align: center;
                    display: block;
                    margin-bottom: 50px;
                }
                @media only screen and (min-width: 1024px) {
                    .loadMore {
                        text-align: left;
                        text-indent: 100px;
                    }
                }
            `}</style>
        </>
    );
};

NewsList.propTypes = {
    newsListData: PropTypes.array,
};

export default NewsList;
