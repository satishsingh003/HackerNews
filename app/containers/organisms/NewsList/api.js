import fetch from '../../../utils/serviceUtil';
import { HACKERNEWS_ENDPOINT } from '../../../constants';

const getMoreNews = async pageNumber => {
  return fetch(`${HACKERNEWS_ENDPOINT}?page=${pageNumber}`)
    .then(response => response.data)
    .catch(err => {
      throw err;
    });
};

export default getMoreNews;
