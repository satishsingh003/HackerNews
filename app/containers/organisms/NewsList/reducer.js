export const APPEND_ITEMS = 'APEND_ITEMS';
export const UPVOTE_ITEM = 'UPVOTE_ITEM';
export const HIDE_ITEM = 'HIDE_ITEM';

export default (state, { type, payload }) => {
  switch (type) {
    case UPVOTE_ITEM: {
      const newsList = state.hits.map(newsItem => {
        if (newsItem.objectID === payload) {
          return {
            ...newsItem,
            upvoted: true,
          };
        }
        return newsItem;
      });
      return { ...state, hits: newsList };
    }

    case HIDE_ITEM: {
      const hits = state.hits.map(newsItem => {
        if (newsItem.objectID === payload) {
          return {
            ...newsItem,
            hidden: true,
          };
        }
        return newsItem;
      });
      return { ...state, hits };
    }

    case APPEND_ITEMS: {
      const mergedNewsItems = [...state.hits, ...payload.hits];
      return { ...state, hits: mergedNewsItems, pageDisplayed: payload.page };
    }

    default:
      return state;
  }
};
