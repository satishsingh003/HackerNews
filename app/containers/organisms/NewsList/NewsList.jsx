import React, { useReducer } from 'react';
import reducer, {APPEND_ITEMS, UPVOTE_ITEM, HIDE_ITEM} from './reducer';
import fetchNews from './api';

import NewsListComponent from '../../../components/organisms/NewsList';

const NewsList = props => {
    const [state, dispatch] = useReducer(reducer, 
        { 
            hits: props.newsListData.hits, 
            pageDisplayed: props.newsListData.page
        }
    );

    const upVoteNewsItem = (newsItem) => {
        dispatch({ type: UPVOTE_ITEM, payload: newsItem.objectID });
    }
    
    const hideNewsItem = (newsItem) => {
        dispatch({ type: HIDE_ITEM, payload: newsItem.objectID });
    }
    
    const loadMoreNews = async () => {
        const moreData = await fetchNews(parseInt(state.pageDisplayed) + 1);
        dispatch({ type: APPEND_ITEMS, payload: moreData });
    }

    return (
        <NewsListComponent
            newsListData={state.hits}
            loadMoreNews={loadMoreNews}
            upVoteNewsItem={upVoteNewsItem}
            hideNewsItem={hideNewsItem}
     />)
};

export default NewsList;
