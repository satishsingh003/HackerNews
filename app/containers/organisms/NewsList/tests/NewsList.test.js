import React from 'react';
import { shallow } from 'enzyme';
import NewsList from '../NewsList';

const propData = {
  hits: [
    {
      objectID: 1,
      title: 'ABC',
    },
    {
      objectID: 12,
      title: 'XYZ',
    },
  ],
  page: 0,
};

describe('testing NewsList container component', () => {
  test('should render correctly', () => {
    const NewsListWrapper = shallow(<NewsList newsListData={propData} />);
    expect(NewsListWrapper).toMatchSnapshot();
  });
});
