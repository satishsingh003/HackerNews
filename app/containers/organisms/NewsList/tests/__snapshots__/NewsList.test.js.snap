// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing NewsList container component should render correctly 1`] = `
<NewsList
  hideNewsItem={[Function]}
  loadMoreNews={[Function]}
  newsListData={
    Array [
      Object {
        "objectID": 1,
        "title": "ABC",
      },
      Object {
        "objectID": 12,
        "title": "XYZ",
      },
    ]
  }
  upVoteNewsItem={[Function]}
/>
`;
